lib/tram/page.rb

```diff
@@ -3,41 +3,35 @@
 class Tram::Page
   extend ::Dry::Initializer
 
+  require_relative "page/section"
+
   class << self
     attr_accessor :i18n_scope
 
     def section(name, options = {})
-      @__sections ||= []
-
-      n = name.to_sym
-      if @__sections.map(&:first).include?(n)
-        raise "Section #{n} already exists"
-      end
+      name = name.to_sym
+      raise "Section #{name} already exists" if sections.key? name
 
-      n = define_section_method(n, options)
-      @__sections << [n, options]
+      section = Section.new(name, options)
+      sections[name] = section
+      define_method(name, &section.block) if section.block
     end
 
     def url_helper(name)
       raise "Rails url_helpers module is not defined" unless defined?(Rails)
       delegate name, to: :"Rails.application.routes.url_helpers"
     end
 
-    private
-
-    def define_section_method(n, options)
-      return n unless options[:value]
-      define_method(n) { instance_exec(&options[:value]) }
-      n
+    def sections
+      @sections ||= {}
     end
   end
 
-  def to_h(options = {})
-    data = page_methods(options).map do |(name, opts)|
-      value = public_send(opts[:method] || name)
-      [name, value]
-    end
-    Hash[data]
+  def to_h(except: nil, only: nil, **)
+    obj = self.class.sections.values.map { |s| s.call(self) }.reduce({}, :merge)
```

Comment:
I think it is somewhat not optimal to calculate all values and then throw out some or most of them.

Reply:
Ahh! Yes, you are right
It would be better to filter out sections, not calculated values